esp/services/ldapenvironment/ldapenvironmentService.cpp

```diff
@@ -152,49 +168,81 @@ bool CldapenvironmentEx::onLDAPCreateEnvironment(IEspContext &context, IEspLDAPC
         formatOUname(respResourcesBaseDN, req.getEnvName(), req.getResourcesMode(), sharedResourcesBaseDN.str(), req.getCustomResourcesBaseDN(), "ou=smc,ou=espservices");
         formatOUname(respWorkunitsBaseDN, req.getEnvName(), req.getWorkunitsMode(), sharedWorkunitsBaseDN.str(), req.getCustomWorkunitsBaseDN(), "ou=workunits");
 
-        // Create HPCC Admin Username
-
-        VStringBuffer respAdminUser("admin_%s", req.getEnvName());
-
+        //Ensure environment doesn't already exist
+        VStringBuffer envOU("ou=%s,%s", req.getEnvName(), ldapRootOU.str());
+        if (secmgr->organizationalUnitExists(envOU.str()))
+        {
+            throw MakeStringException(-1, "Environment '%s' already exists, please specify a different name", envOU.str());
+        }
 
         //Create LDAP resources OU Hierarchy
 
         if (req.getCreateLDAPEnvironment())
         {
             VStringBuffer description("%s, created by %s", req.getEnvDescription(), req.getEnvOwnerName());
+            //Note that ESP will also try to create these OUs on startup
             try { secmgr->createLdapBasedn(nullptr, respFilesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) {}
+            catch(...) { notes.appendf("\nError creating '%s'", respFilesBaseDN.str()); }
             try { secmgr->createLdapBasedn(nullptr, respGroupsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) {}
+            catch(...) { notes.appendf("\nError creating '%s'", respGroupsBaseDN.str()); }
             try { secmgr->createLdapBasedn(nullptr, respUsersBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) {}
+            catch(...) { notes.appendf("\nError creating '%s'", respUsersBaseDN.str()); }
             try { secmgr->createLdapBasedn(nullptr, respResourcesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) {}
+            catch(...) { notes.appendf("\nError creating '%s'", respResourcesBaseDN.str()); }
             try { secmgr->createLdapBasedn(nullptr, respWorkunitsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
-            catch(...) {}
-
-            // NOTE!  HPCC Admin group and HPCC Admin user will be
-            // created when ECLWatch is started in Admin mode
+            catch(...) { notes.appendf("\nError creating '%s'", respWorkunitsBaseDN.str()); }
         }
         
+        //----------------------------------
+        // Create HPCC Admin Username/password.
+        // Attempt to create the Kubernetes secret for that user
+        // NOTE!  HPCCAdmin LDAP group and HPCCAdmin LDAP user will be
+        // created by ESP when started in Admin mode
+        //----------------------------------        
+        VStringBuffer respHPCCAdminUser("admin_%s", req.getEnvName());
+        StringBuffer  respHPCCAdminPwd;
+        StringBuffer  hpccAdminK8sSecretName("hpccadminsecret");
```

Comment:
I would add the environment name (envName) to the k8s secret name so there aren't collisions as people use the service to create multiple environments over time.

Reply:
Done